### PR TITLE
feat: support P2E bank DiffTest

### PR DESCRIPTION
**Pull request by daiyongyuan**
## Summary

- update Bebop and bbdev submodules for P2E Bank DiffTest
- allow a full private-bank mapping table to reallocate the physical bank already owned by the target vbank
- route smatmul writes through tracked bank indexing so BEMU initialization canonicalization remains accurate

## Dependencies

- https://github.com/DangoSys/bebop/pull/21
- https://github.com/DangoSys/bbdev/pull/17

## Validation

- all repository pre-commit hooks passed, including scalafmt
- child repository unit tests passed
- P2E LeNet hardware DiffTest: pass=631 mismatch=0 missing_rtl=0 unexpected_rtl=0

The RTL allocation change requires a rebuilt bitstream to exercise the full-bank reallocation edge case. P2E DiffTest itself reuses an existing bitstream.

### CODE REVIEW

File: .github/scripts/ci_repo_lock.sh [hunk 1]
```diff
@@ -31,7 +31,7 @@ repo_reset() {
   rm -rf compiler/thirdparty/buddy-mlir/llvm/build
   rm -rf compiler/thirdparty/buddy-mlir/build
   git submodule sync
-  git submodule update --init --force -- $(git ls-tree -r "${want_sha}" | awk '$1 == "160000" { print $4 }')
+  git submodule update --init --force --recursive -- $(git ls-tree -r "${want_sha}" | awk '$1 == "160000" { print $4 }')
```
Comment (shirohasuki) [inline .github/scripts/ci_repo_lock.sh]: We don't allow `--recursive`. Are you sure this will not cause Chipyard to download recursively? There may be no permission for some of the submodules, in my experience.

**Review by shirohasuki — CHANGES_REQUESTED** [review]

File: .github/scripts/ci_repo_lock.sh (hunk 1 above)
Comment (daiyongyuan) [reply .github/scripts/ci_repo_lock.sh]: Good catch. I removed `--recursive` in 129ed1e8. The first update initializes only the root-level gitlinks from the target tree; the second explicitly initializes only `compiler/thirdparty/buddy-mlir/llvm`, which is the nested source missing in CI. Chipyard nested submodules are not traversed. Local pre-commit and the toy compiler build pass.

Comment (daiyongyuan) [reply .github/scripts/ci_repo_lock.sh]: One correction to my previous reply: Git does not recognize the nested path when the update is invoked from the root repository. I fixed that in bec6c908 by running `git -C compiler/thirdparty/buddy-mlir submodule sync` and `git -C compiler/thirdparty/buddy-mlir submodule update --init --force -- llvm`. This remains non-recursive and does not traverse Chipyard.

File: arch/src/main/scala/framework/frontend/Frontend.scala [hunk 2]
```diff
@@ -17,12 +19,15 @@ import framework.balldomain.blink.SubRobRow
 @instantiable
 class Frontend(val b: GlobalConfig) extends Module {
 
+  val sharedHashCount = if (b.memDomain.sharedEnable) SharedMemLayout.totalBank(b) else 0
+
   @public
   val io = IO(new Bundle {
+    val hartid = Input(UInt(b.cpu.xLen.W))
```
Comment (shirohasuki) [inline arch/src/main/scala/framework/frontend/Frontend.scala]:
```suggestion
    val hartid = Input(UInt(b.tile.xLen.W))
```

File: arch/src/main/scala/framework/frontend/Frontend.scala [hunk 3]
```diff
@@ -17,12 +19,15 @@ import framework.balldomain.blink.SubRobRow
...
 
     // RoCC command input
     val cmd = Flipped(Decoupled(new Bundle {
-      val cmd = new RoCCCommandBB(b.core.xLen)
+      val cmd = new RoCCCommandBB(b.cpu.xLen)
```
Comment (shirohasuki) [inline arch/src/main/scala/framework/frontend/Frontend.scala]:
```suggestion
      val cmd = new RoCCCommandBB(b.tile.xLen)
```

File: arch/src/main/scala/framework/frontend/Frontend.scala [hunk 4]
```diff
@@ -36,9 +41,17 @@ class Frontend(val b: GlobalConfig) extends Module {
...
+    val bank_hashes =
+      if (b.sim.diffTest) {
+        Some(Input(Vec(b.memDomain.bankNum + sharedHashCount, new PhysicalBankHash(b))))
+      } else {
+        None
+      }
 
     // RoCC response
-    val resp    = Decoupled(new RoCCResponseBB(b.core.xLen))
+    val resp    = Decoupled(new RoCCResponseBB(b.cpu.xLen))
```
Comment (shirohasuki) [inline arch/src/main/scala/framework/frontend/Frontend.scala]:
```suggestion
    val resp    = Decoupled(new RoCCResponseBB(b.tile.xLen))
```

File: bebop [hunk 5]
```diff
@@ -1 +1 @@
-Subproject commit 490c9aa942a9425727ab7db37eb67075ba234149
+Subproject commit f3550acb282d49ab97e724e89c98f074f478eaa2
```
Comment (shirohasuki) [inline bebop]:
```suggestion
Subproject commit f3550acb282d49ab97e724e89c98f074f478eaa2-dirty
```
